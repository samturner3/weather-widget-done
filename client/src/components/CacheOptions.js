import React from 'react';
import moment from 'moment';
import update from 'immutability-helper';

export default class CacheOptions extends React.Component {
	constructor(props) {
		super(props);
		this.state = {
			cacheObjects: {}
		};
	}

	componentWillMount = () => {
		const initialObj = {};
		this.props.placesCache.forEach((cache, index) => {
			const time = moment(cache.exp).fromNow();
			const id = `${cache.key}`;
			initialObj[id] = {
				time
			}
		});
		this.setState({ cacheObjects: initialObj });
	};

	componentDidMount = () => {
		this.interval = setInterval(() => {
			this.props.placesCache.forEach((cache, index) => { // Ensure there is a state for this prop, and update it.
				this.updateTime(cache, index);
			});
		}, 1000);
	};

	updateTime = (cache, index) => {
		const time = `${moment(cache.exp).fromNow()}`;
		const lookup = `${cache.key}`;
		if (this.state.cacheObjects[lookup] === undefined) {
			this.setState(({
				cacheObjects: update(this.state.cacheObjects, {[`${cache.key}`]: {$set: {time}}})
			}));
		} else {
			this.setState(({
				cacheObjects: update(this.state.cacheObjects, {[`${cache.key}`]: {$set: {time}}})
			}));
		}
	}
	render() {
		return (
			<ul>
				{this.props.placesCache.map((cache, index) => {
					if ((moment().diff(cache.exp, 'seconds') < 0) && this.state.cacheObjects[`${cache.key}`] !== undefined) {
						return (<li key={index}>{cache.name} <i>(Expires {this.state.cacheObjects[`${cache.key}`].time}, {moment().diff(cache.exp, 'seconds')})</i></li>)
					}
				})}
			</ul>
		)
	}
};
