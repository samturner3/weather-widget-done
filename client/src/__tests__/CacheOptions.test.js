import React from 'react';
import CacheOptions from '../components/CacheOptions';

jest.useFakeTimers();
const inTenMins = Date.now() + 600000;

function setup(exp = inTenMins) {
	const props = {
		placesCache: [
			{
				exp,
				key: '-33.8688197+151.20929550000005',
				name: 'Sydney'
			}
		]
	};
	const wrapper = shallow(<CacheOptions {...props} />);

	return {
		wrapper,
		props
	};
}

describe('CacheOptions', () => {
	it('should render with right amount of elements', () => {
		const { wrapper } = setup();
		expect(wrapper).toMatchSnapshot();
		expect(wrapper.find('li')).toHaveLength(1);
	});

	it('should call the updateTime method when component has been mounted', () => {
		const { wrapper } = setup();
		const updateTimeSpy = jest.spyOn(wrapper.instance(), 'updateTime');
		// allow the set interval to run just once
		jest.runTimersToTime(1000);
		expect(updateTimeSpy).toHaveBeenCalledTimes(1);
	})

	it('should render no <li />  element if chache has expired', () => {
		const { wrapper } = setup(1535015102520);
		expect(wrapper.find('li')).toHaveLength(0)
	});

	// TODO: make this test pass
	// describe('updateTime', () => {
	// 	it('ss', () => {
	// 		const { wrapper, props } = setup();
	// 		const cache = Object.assign({}, props.placesCache[0]);
	// 		cache.key = 'not there'
	// 		wrapper.instance.updateTime(cache);
	// 		// console.log(cache)
	// 	})
	// });
})
