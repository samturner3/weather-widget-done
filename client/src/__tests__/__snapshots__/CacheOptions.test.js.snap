// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CacheOptions should render with right amount of elements 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <CacheOptions
    placesCache={
        Array [
            Object {
              "exp": 1535070680000,
              "key": "-33.8688197+151.20929550000005",
              "name": "Sydney",
            },
          ]
    }
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <li>
          Sydney
           
          <i>
                    (Expires 
                    in 10 minutes
                    , 
                    -600
                    )
          </i>
</li>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "0",
        "nodeType": "host",
        "props": Object {
          "children": Array [
            "Sydney",
            " ",
            <i>
              (Expires 
              in 10 minutes
              , 
              -600
              )
</i>,
          ],
        },
        "ref": null,
        "rendered": Array [
          "Sydney",
          " ",
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                "(Expires ",
                "in 10 minutes",
                ", ",
                -600,
                ")",
              ],
            },
            "ref": null,
            "rendered": Array [
              "(Expires ",
              "in 10 minutes",
              ", ",
              -600,
              ")",
            ],
            "type": "i",
          },
        ],
        "type": "li",
      },
    ],
    "type": "ul",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <li>
            Sydney
             
            <i>
                        (Expires 
                        in 10 minutes
                        , 
                        -600
                        )
            </i>
</li>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "0",
          "nodeType": "host",
          "props": Object {
            "children": Array [
              "Sydney",
              " ",
              <i>
                (Expires 
                in 10 minutes
                , 
                -600
                )
</i>,
            ],
          },
          "ref": null,
          "rendered": Array [
            "Sydney",
            " ",
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  "(Expires ",
                  "in 10 minutes",
                  ", ",
                  -600,
                  ")",
                ],
              },
              "ref": null,
              "rendered": Array [
                "(Expires ",
                "in 10 minutes",
                ", ",
                -600,
                ")",
              ],
              "type": "i",
            },
          ],
          "type": "li",
        },
      ],
      "type": "ul",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getSnapshotBeforeUpdate": true,
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
